prebid-server/pbs-modules/rules-engine.md

```diff
@@ -0,0 +1,805 @@
+- removes bidderD in the EU datacenter if that bidder hasn't synced with this user
+- removes bidderB and bidderF in the APAC datacenter
+- logs any bidder removed to analytics
+
+2% of the time
+
+- logs "control" to analytics
+
+```json
+{
+  "hooks": {
+    "modules": {
+      "pb-rules-engine": {
+        "enabled": true,
+        "timestamp": "20250314 00:00:00",
+        "ruleSets": [
+          {
+            "stage": "processed-auction-request",
+            "name": "remove-bidder-by-datacenter",
+            "version": "1234",
+            "modelGroups": [{
+                "weight": 98,
+                "analyticsKey": "rm-bidder-by-dc",
+                "version": "4567",
+                "schema": [{"function": "datacenters"}],
+                "default": [],
+                "rules": [{
+                    "conditions": ["eu"],
+                    "results": [{
+                        "function": "excludeBidders",
+                        "args": [{
+                            "bidders": [ "bidderA"],
+                            "seatnonbid": 203,
+                            "analyticsValue": "rm-eu"
+                          },{
+                            "bidders": ["bidderD"],
+                            "seatnonbid": 203,
+                            "ifSyncedId": false,
+                            "analyticsValue": "rm-eu-nosync"
+                          }
+                        ]
+                      }
+                    ]
+                  },{
+                    "conditions": ["apac"],
+                    "results": [{
+                        "function": "excludeBidders",
+                        "args": [{
+                            "bidders": ["bidderB", "bidderF"],
+                            "seatnonbid": 203,
+                            "analyticsValue": "rm-apac"
+                          }
+                        ]
+                      }
+                    ]
+                  }
+                ]
+              },{
+                "weight": 2,
+                "analyticsKey": "rm-bidder-by-dc",
+                "default": [
+                  {
+                    "function": "logAtag",
+                    "args": {
+                      "analyticsValue": "control"
+                    }
+                  }
+                ]
+
+              }
+            ]
+          }
+        ]
+      }
+    }
+  }
+}
+```
+
+## How it Works
+
+Similar to the Floors feature, it's assumed that an offline machine-learning algorithm will be part of a bidder optimization feedback loop to adjust to changing conditions. Here's the assumed architecture:
+
+![Prebid Server Rules Engine Module](/assets/images/prebid-server/pbs_rules_engine_context.png){:class="pb-xlg-img"}
+
+1. Auction requests come in that trigger the Rules Engine Module.
+1. The module reads the current set of rules from account configuration or by directly fetching them from the service provider.
```

Comment:
Should we add another bullet here or is this assumed? "The results of the optimization decision are applied to the auction"

Reply:
Added